staging/src/k8s.io/apiserver/pkg/server/filters/waitgroup.go

```diff
@@ -0,0 +1,53 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"net/http"
+	"sync"
+
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+)
+
+// WithWaitGroup adds all non long-running requests to wait group, which is used for graceful shutdown.
+func WithWaitGroup(handler http.Handler, requestContextMapper apirequest.RequestContextMapper, longRunning apirequest.LongRunningRequestCheck, wg *sync.WaitGroup) http.Handler {
+	if longRunning == nil {
+		return handler
```

Comment:
This is surprising. I would expect that we handle all request if longRunning is nil, i.e. we assume `longRunning == false` if none is given.

Reply:
If we handle all request, then the waitGroup.wait will block forever. Because longRunning  request handler may not finish.

Reply:
Should we warn then at least? this looks like a programming error using WithWaitGroup, but longRunning==nil.

Reply:
OK, should warning

Reply:
```
	genericConfig.LongRunningFunc = filters.BasicLongRunningRequestCheck(
		sets.NewString("watch", "proxy"),
		sets.NewString("attach", "exec", "proxy", "log", "portforward"),
	)
```
This code gurantee longRunning != nil, I think this if branch can be removed.